test/unit/authorizer_test.rb

```diff
@@ -10,169 +10,207 @@ def setup
+            assert_equal [domain4], collection
+            collection = auth.find_collection(Domain, :permission => :delete_domains)
+            assert_equal [], collection
+            collection = auth.find_collection(Domain)
+            assert_equal [domain2, domain3, domain4], collection
+
+            refute auth.can?(:view_domains, domain1, cache)
+            assert auth.can?(:view_domains, domain2, cache)
+            assert auth.can?(:view_domains, domain3, cache)
+            refute auth.can?(:view_domains, domain4, cache)
+            refute auth.can?(:edit_domains, domain1, cache)
+            refute auth.can?(:edit_domains, domain2, cache)
+            refute auth.can?(:edit_domains, domain3, cache)
+            assert auth.can?(:edit_domains, domain4, cache)
+
+            # unlimited filter on Domain permission does add the domain
+            FactoryGirl.create(:filter, :role => @role, :permissions => [permission1])
+            collection = auth.find_collection(Domain)
+            assert_includes collection, domain1
+            assert_includes collection, domain2
+            assert_includes collection, domain3
+            assert_includes collection, domain4
+          end
+
+          test "for user without filter" do
+            permission = Permission.find_by_name('view_domains')
+            FactoryGirl.create(:filter, :role => @role, :permissions => [permission])
+            domain     = FactoryGirl.create(:domain)
+            auth       = Authorizer.new(FactoryGirl.create(:user))
+
+            result = auth.find_collection(Domain, :permission => :view_domains)
+            assert_not_includes result, domain
+            assert_kind_of ActiveRecord::Relation, result
+            refute auth.can?(:view_domains, domain, cache)
+          end
+
+          test "caches results per permission and class" do
+            permission1 = Permission.find_by_name('view_domains')
+            FactoryGirl.create(:filter, :on_name_starting_with_a,
+                               :role => @role, :permissions => [permission1])
+            domain1     = FactoryGirl.create(:domain, :name => 'a-domain.to-be-found.com')
+            domain2     = FactoryGirl.create(:domain, :name => 'x-domain.not-to-be-found.com')
+            permission2 = Permission.find_by_name('view_architectures')
+            architecture = FactoryGirl.create(:architecture)
+            FactoryGirl.create(:filter, :role => @role, :permissions => [permission2])
+
+            auth = Authorizer.new(@user)
+
+            domains_collection = [domain1]
+            domains_collection.stubs(:where).
+              with(:id => domain1.id).returns([domain1])
+            domains_collection.stubs(:where).
+              with(:id => domain2.id).returns([])
+            domains_collection.stubs(:where).
+              with(:id => architecture.id).returns([])
+            auth.stubs(:find_collection).returns(domains_collection).at_most(6)
```

Comment:
Using `at_most` here is going to stop this testing that the cache works when `cache=true` because it'll make 6 calls like `cache=false`.  Perhaps go back to `times` by doing: `times(cache ? 3 : 6)`?